src/substrait/function_registry.py

```diff
@@ -0,0 +1,316 @@
+from substrait.gen.proto.parameterized_types_pb2 import ParameterizedType
+from substrait.gen.proto.type_pb2 import Type
+from importlib.resources import files as importlib_files
+import itertools
+from collections import defaultdict
+from collections.abc import Mapping
+from pathlib import Path
+from typing import Any, Optional, Union
+from .derivation_expression import evaluate
+
+import yaml
+import re
+
+_normalized_key_names = {
+    "binary": "vbin",
+    "interval_compound": "icompound",
+    "interval_day": "iday",
+    "interval_year": "iyear",
+    "string": "str",
+    "timestamp": "ts",
+    "timestamp_tz": "tstz",
+}
+
+
+def normalize_substrait_type_names(typ: str) -> str:
+    # First strip off any punctuation
```

Comment:
Question marks are the nullability for a given type.  I'd call that out explicitly.